[no_early_kickoff][Core][Dashboard]Configurable dashboard RPC port and public IP

joncarter1 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This PR achieves the following:
1. Makes the dashboard head process' RPC port configurable using the `--dashboard-grpc-port` CLI flag or `RAY_DASHBOARD_RPC_PORT` env var.
2. Makes the dashboard's public IP address (put in the GCS) use `--node-ip-address` from the head node if set. This is necessary because the `DASHBOARD_RPC_ADDRESS` put into the GCS currently uses `ray.util.get_node_ip_address()` even if `--node-ip-address` is set. This does not work in Docker containers with an isolated network as the function will return the container IP.

Together these changes allow Ray to work in Docker containers which are isolated from the host network (closes #33534 )

## Related issue number

This PR achieves the goal of #33534 , which is to provide full Docker isolation from the host network.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

rkooo567 commented:
Let me take a look at it today!

jjyao commented on dashboard/head.py:
```diff
@@ -17,7 +17,11 @@
 from ray.dashboard.utils import async_loop_forever
-from ray.dashboard.consts import DASHBOARD_METRIC_PORT
+from ray.dashboard.consts import (
+    DASHBOARD_METRIC_PORT,
```
@rkooo567 what's the difference between `DASHBOARD_METRIC_PORT` and `--metrics-export-port`?

joncarter1 replied:
@jjyao if I understand correctly, DASHBOARD_METRIC_PORT is specifically for exposing Dashboard metrics to Prometheus. `--metrics-export-port` exposes more general metrics from the node (and is common to the head and workers).

jjyao reviewed (commented):
Thanks for the contribution.

I feel we should configure ports in a consistent way: always use command line options, instead of some are env vars and some are command line options.

rkooo567 commented:
@jjyao do you see any downside of supporting both ways? I think it is not a bad idea to support both. It is too late to not allowing CLI based port configuration, so if we have to be consistent, we should use CLI.

joncarter1 commented:
> For this PR, can you also add the CLI flag? And update the documentation (configuring ray page).

Would you like me to add a CLI flag for the `DASHBOARD_METRICS_PORT` too?

joncarter1 commented:
@rkooo567 @jjyao after investigating what it would take to implement a CLI flag for the port I am opposed to the idea. The process of figuring out what it would take did a good job of highlighting a need for a broader refactor of the way Ray ports are configured. 

Below, I will try and illustrate exactly what the process of adding CLI support would look like under the current paradigm, following the existing convention for CLI-supported ports.

First up, the port must be added to the `ray start` CLI:
1. Add CLI flag to the click parameters and the `start` function arguments: https://github.com/ray-project/ray/blob/ee271123ea3d8518fd7fdb08f25085e9ced71e57/python/ray/scripts/scripts.py#L529-L530
2. Next up, add the port to the RayParams object initialisation:
https://github.com/ray-project/ray/blob/ee271123ea3d8518fd7fdb08f25085e9ced71e57/python/ray/_private/parameter.py#L12
This class is pretty bloated, and there are methods e.g. https://github.com/ray-project/ray/blob/ee271123ea3d8518fd7fdb08f25085e9ced71e57/python/ray/_private/parameter.py#L284
that would also need to be updated with the new port to stay aligned with the existing convention.
3. Add the port to the [Node class](https://github.com/ray-project/ray/blob/master/python/ray/_private/node.py) (init+attribute). This class is also quite bloated with a lot of data clump. Again, if we're to stay with convention, there are a number of places where the new port would need to be added e.g. adding it as a property.

The dashboard is then started by `start_api_server`:
https://github.com/ray-project/ray/blob/ee271123ea3d8518fd7fdb08f25085e9ced71e57/python/ray/_private/services.py#L1040
So:
4. The port from the node class would need to be added as an argument to this function.
This function then calls the ray dashboard executable, so this means:
5. The port needs to be added to parseargs CLI flags and Dashboard init here: https://github.com/ray-project/ray/blob/master/dashboard/dashboard.py
6. The port needs to be propagated through the Dashboard class to the DashboardHead class to where it is finally used.

That's _a lot_ of places where code needs to change just to make one port configurable...

I hope I've made the case for an environment-based approach on this!

This PR already adds to the docs that the new port and the existing `DASHBOARD_METRICS_PORT` can be configured from the environment. I can also add a short section on the new `RAY_DASHBOARD_IP_ADDRESS` variable introduced to set the reachable IP of the dashboard when running in a network-isolated container.

Please let me know if that's okay with you both.

jjyao commented on dashboard/consts.py:
```diff
@@ -25,6 +25,9 @@
 PURGE_DATA_INTERVAL_SECONDS = 60 * 10
 ORGANIZE_DATA_INTERVAL_SECONDS = 2
 DASHBOARD_RPC_ADDRESS = "dashboard_rpc"
+DASHBOARD_RPC_PORT = env_integer("DASHBOARD_RPC_PORT", 0)
```
Let's prefix the env var with `RAY_`?

jjyao commented:
@joncarter1,

I completely understand the work to propagating the cli option to the right place. 

I can see pros and cons from both side:

1. cli gives you nice `ray start --help`
2. env var is easier to use

rkooo567 commented:
It's a little unfortunate, but I don't think we can churn cli based approach. It's been used pretty long time, and I assume most of users already use this port configuration CLI. Standardize users to use env var seems like too big public API change for minor releases.  

If we really need an env var approach, I think the best we can do is just to support both and slowly recommend env var based approach if we'd like to change.

joncarter1 commented on python/ray/tests/test_dashboard.py:
```diff
@@ -195,16 +194,38 @@ def f():
     with pytest.raises(
         ray.exceptions.RuntimeEnvSetupError,
-        match="the grpc service of agent is invalid",
+        match="Ray agent couldn't be started due to the port conflict",
```
I updated this test because it seemed to be failing independently of my changes.
(I can't find anything in the code base that matches the original regex match)

I can see it is skipped on Windows, does it usually get run in CI pipelines?

joncarter1 commented:
@rkooo567 @jjyao I have added CLI support using the new `--dashboard-head-grpc-port` flag.

I have removed the previously introduced `RAY_DASHBOARD_IP_ADDRESS` env var and instead propagated the `--node-ip-address` parameter through to the dashboard. I believe this should be the intended behavior, rather than the current behavior of the dashboard head always calling `ray.util.get_node_ip_address()` to determine it's publicly reachable IP even when `--node-ip-address` is specified.

This means the `--node-ip-address` flag works correctly in network-isolated Docker containers.

joncarter1 commented:
@rkooo567 I have fixed the tests that were explicitly failing due to changes introduced in this PR.

`workflow-tests` looks to be failing because it doesn't perform a full build. I expect it should pass once this PR goes into the nightlies.

joncarter1 commented on dashboard/dashboard.py:
```diff
@@ -88,6 +94,19 @@ async def run(self):
+    )
+    parser.add_argument(
+        "--dashboard-ip",
+        required=False,
```
Sounds good.

joncarter1 commented on dashboard/dashboard.py:
```diff
@@ -88,6 +94,19 @@ async def run(self):
+        help="The port for the dashboard to listen for gRPC on.",
+    )
+    parser.add_argument(
+        "--dashboard-ip",
```
I had named this to be consistent with the `monitor-ip` flag used here:
https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/_private/monitor.py
(as a similar example of a process being spawned by the head node.)

Happy to change though.

joncarter1 commented on dashboard/dashboard.py:
```diff
@@ -88,6 +94,19 @@ async def run(self):
         "--gcs-address", required=True, type=str, help="The address (ip:port) of GCS."
     )
+    parser.add_argument(
+        "--dashboard-head-grpc-port",
```
SGTM

I thought it best to keep the CLI flag being propagated from `ray start` to the dashboard executable the same but I've already broken that rule with `--node-ip-address` -> `--dashboard-ip` 😅

joncarter1 commented on python/ray/scripts/scripts.py:
```diff
@@ -430,6 +430,12 @@ def debug(address):
     help="the port for dashboard agents to listen for grpc on.",
 )
+@click.option(
+    "--dashboard-head-grpc-port",
```
Given the existing dashboard-related port flags:
"--dashboard-port", "--dashboard-agent-listen-port", "--dashboard-agent-grpc-port"
I'd tried to be consistent with the latter two.

Happy to change.

rkooo567 replied:
Let's just call it `dashboard-grpc-port`. I think api-server is not exposed yet

joncarter1 commented:
oh dear. I tried to rebase on upstream/master to fix a test and that has caused complete chaos...

rkooo567 reviewed (approved):
LGTM! Just one last request for tests!

joncarter1 commented on python/ray/tests/test_dashboard.py:
```diff
@@ -195,16 +194,35 @@ def f():
+    address = call_ray_start
+    addresses = ray.init(address=address)
+    dashboard_url = addresses["webui_url"]
+    # Dashboard should fail to start with conflicting head port
```
@rkooo567 I've added another test which checks dashboard starts with a non-conflicting, configured gRPC port, was that what you meant?

joncarter1 commented on python/ray/tests/test_dashboard.py:
```diff
@@ -195,16 +195,50 @@ def f():
+    assert addresses.dashboard_url == "127.0.0.1:8265"
+
+
+@pytest.mark.parametrize(
+    "listen_port",
+    [conflict_port],
+    indirect=True,
+)
+def test_dashboard_grpc_port_conflict(listen_port, call_ray_stop_only, shutdown_only):
+    try:
+        subprocess.check_output(
+            [
+                "ray",
+                "start",
+                "--head",
+                "--dashboard-grpc-port",
+                f"{conflict_port}",
+                "--include-dashboard=True",
+            ],
+            stderr=subprocess.PIPE,
+        )
+    except subprocess.CalledProcessError as e:
+        assert f"Failed to bind to address 0.0.0.0:{conflict_port}".encode() in e.stderr
```
@rkooo567 have updated the test accordingly

jjyao commented on python/ray/scripts/scripts.py:
```diff
@@ -430,6 +430,12 @@ def debug(address):
     help="the port for dashboard agents to listen for grpc on.",
 )
+@click.option(
+    "--dashboard-grpc-port",
```
ideally, I think `--dashboard-rpc-port` is better given grpc is just implementation details but since we already have `--dashboard-agent-grpc-port`, I think it's fine to name this way.

jjyao reviewed (approved):
LGTM. I'll leave to @rkooo567 for merge.

Also, thanks a bunch for doing this!!!

joncarter1 commented:
You're both very welcome! Sorry about the Git carnage! :sweat_smile:

rkooo567 commented:
test_dashboard -> this failure seems related! Under the test_dashboard

joncarter1 commented:
> test_dashboard -> this failure seems related! Under the test_dashboard

Good spot, test fixed.

joncarter1 commented:
@rkooo567 is there anything else you'd like me to address?
The remaining failures appear to be caused by not doing a full build of Ray which _should_ be fixed when this goes into the nightlies. 

This could be resolved by making the `--node-ip-address` optional within the dashboard executable if there's any concern about impact on other MRs.

jjyao commented:
@joncarter1 could you rebase with master?

joncarter1 commented:
@jjyao done :)

jjyao commented:
Failed tests are unrelated.

jjyao commented:
Thanks again for the great contribution! Looking forward to more of them :)
